runtime/gc_structs/ClassLoaderClassesIterator.cpp

```diff
@@ -99,6 +102,52 @@ GC_ClassLoaderClassesIterator::switchToSystemMode()
 	return isSystemClassLoader;
 }
 
+J9Class *
+GC_ClassLoaderClassesIterator::nextArrayClass()
+{
+	while (_arrayState != state_done) {
+		switch(_arrayState) {
+			case state_valueTypeArray:
+#if defined(J9VM_OPT_VALHALLA_FLATTENABLE_VALUE_TYPES)
+				if (NULL != J9CLASS_GET_NULLRESTRICTED_ARRAY(_startingClass)) {
+					_iterateArrayClazz = J9CLASS_GET_NULLRESTRICTED_ARRAY(_startingClass);
+					_arrayState = state_valueTypeArrayList;
+				} else
+#endif /* defined(J9VM_OPT_VALHALLA_FLATTENABLE_VALUE_TYPES) */
+				if (NULL != _startingClass->arrayClass) {
+					_iterateArrayClazz = _startingClass->arrayClass;
+					_arrayState = state_arrayList;
+				} else {
+					_iterateArrayClazz = NULL;
+					_arrayState = state_done;
+				}
+				break;
+			case state_valueTypeArrayList:
+				if (NULL == _iterateArrayClazz) {
+					_iterateArrayClazz = _startingClass->arrayClass;
+					_arrayState = state_arrayList;
+				} else {
+					_iterateArrayClazz = _iterateArrayClazz->arrayClass;
+				}
+				break;
+			case state_arrayList:
+				if (NULL == _iterateArrayClazz) {
+					_arrayState = state_done;
+				} else {
+					_iterateArrayClazz = _iterateArrayClazz->arrayClass;
+				}
+				break;
+			case state_done:
+			default:
+				break;
+		}
+		if (NULL != _iterateArrayClazz) {
+			break;
+		}
+	}
```

Comment:
It is ok as-is, but you can reorganize this code even more. Currently you have two states for NULL restricted - `state_valueTypeArray` where you setup list walk and `state_valueTypeArrayList` where list is walked. Similarly you can have two states for ArrayList too. You do have  `state_arrayList` where list is walked, you can add `state_array` where this walk is setup. In this case  you can move `_iterateArrayClazz = _startingClass->arrayClass;...` there. The reason for this suggestion technically `state_valueTypeArrayList` or `state_valueTypeArray` should not know details related to `state_arrayList`, just know which state is next. I think it helps to avoid code duplication and make logic cleaner.
You can move this code to the new `state_array`:
```
				if (NULL != _startingClass->arrayClass) {
					_iterateArrayClazz = _startingClass->arrayClass;
					_arrayState = state_arrayList;
				} else {
					_iterateArrayClazz = NULL;
					_arrayState = state_done;
				}
```
and reuse in `state_valueTypeArrayList` too.

Reply:
Thanks for the feedback, please check the latest updates to make sure it's what you had in mind.

The first state still knows some array list details `state_valueTypeArray`, this was previously added to avoid unnecessary switches when no array lists exist.
```
				if (NULL != _startingClass->arrayClass) {
					_arrayState = state_array;
				} else {
					_arrayState = state_done;
				}
```

Reply:
Yes, very good, thank you! With simplification for `state_valueTypeArray` state it looks clean and easy to read.

Reply:
If I write this I will use `NULL !=` instead of `NULL ==` and swap halves of if. So, if condition is TRUE continue to iterate in the same state, and do state switch in the `else` part. But this is stylish thing, I am not insisting :-)

Reply:
That looks better to me to, I rearranged the logic.